Deprecation messages (#65)
andycroll/andycroll.com: _posts/2019/2019-10-20-use-a-deprecation-message.md

---
title: "Use Active Support in Rails for deprecation messages"
description: "An internal Rails tool, availible for us"
layout: article
category: ruby
image:
  base: '2019/use-active-support-in-rails-for-deprecation-messages'
  alt: 'Caution on the pavement'
  credit: Goh Rhy Yan
  source: "https://unsplash.com/photos/FFgcWvplwsc"

---

> When authors of open source libraries change major functionality, or are removing features they often adhere to something called [semantic versioning](https://semver.org) and don't remove things until a major version is released.

Reviewer: *or remove features, they often..

Reviewer: Having the '...and don't remove things...' part at the end makes the sentence read strangely, especially as you refer to them removing things earlier on in the sentence.

I think start with changing major functionality = semver, then say - features aren't removed until major versions are released.

> Before the big changes it’s a good idea to add messages to users of the software to mention forthcoming changes in a harmless but visible way. This is often done by sending messages to the logs when code runs.

Reviewer: ```suggestion
Before any big changes it’s a good idea to send messages to users of the software in a harmless, but visible, way. This is often done via the logs output when the code runs.
```

You might have seen these when your tests run or in your live application logs.

```
blah/blah/file.rb:34: warning: constant ::Fixnum is deprecated
```

> Although in our private codebases we’re often not making big releases that could impact hundreds of other unseen developers, it's still a good idea to help those in your team (including yourself) when functionality changes.

Reviewer: This sentence could be more succinct!

> The Rails framework does this enough that it has [a standard depreciation approach baked into Active Support](https://api.rubyonrails.org/classes/ActiveSupport/Deprecation/Reporting.html), and in _our_ code we can piggyback off the feature they use to deprecate internal features.

Reviewer: *goes through this (cycle) enough (?)

## Instead of...

...just logging

```ruby
class Coffee
  def make
    put_instant_in_cup
    add_hot_water
    puts "We should be using an Aeropress"
  end

  def make_properly
    grind_on_medium_fine
    add_to_inverted_aeropress
    sleep(37)
    invert_and_apply_pressure
  end
  # ...

> end

Reviewer: Ha — love this example.

Reviewer: But wouldn't we want the message to be a deprecation warning about `make` to use `make_properly` instead?

```

## Use...

...Active Support’s deprecation library.

```ruby
class Coffee
  def make
    put_instant_in_cup
    add_hot_water
    ActiveSupport::Deprecation.warn("We should be using an Aeropress")
  end

  def make_properly
    grind_on_medium_fine
    add_to_inverted_aeropress
    sleep(37)
    invert_and_apply_pressure
  end
  # ...
end
```

When you of your co-workers call `Coffee.new.make` you’ll see...

```

> DEPRECATION WARNING: We should be using an Aeropress (called from coffee.rb:4)

Reviewer: See above comment re a deprecation warning.

```

... in your logs.

## Why?

> Firstly, instant coffee is horrible and you _just shouldn’t_.

Reviewer: *First

Reviewer: Do we want judgements in the newsletter...? 🤔 (As much as I, and probably the bulk of the readership, will agree with you...)

Author: Damn right. War on terible coffee.

> Secondly, this is a relatively lightweight way to 'soft protect' old ways of doing things in favour of new code.

Reviewer: *Second,

> It's useful in two ways. One as a way to discover places in the code that use your existing method, it might be called in places you don’t expect. Another is to prevent your coworkers from adding new uses of the deprecated method, when they do, they’ll see your message.

Reviewer: But you've already said 'First' and 'Second'. Do you mean the 'Second' point, i.e. the 'soft protect' is useful in two ways?

If so, maybe say 'this' is useful in two ways....

Reviewer: Full-stop before "It might be called in places you don't expect."

Reviewer: *because/since, when they attempt to, they'll see the deprecation message.

> A good example of how you might use it in your application might be to use it in scopes, where you want to maintain existing logic, but discourage use.

Reviewer: ```suggestion
Another way to use Rails' deprecation functionality is within scopes. You'd do this when you want to maintain existing logic, but discourage further use.
```

```ruby
class TakeawayCoffeeCup < ApplicationRecord
  scope :good_choice, -> {
    ActiveSupport::Deprecation.warn("Be more ecologically minded")
    where(material: "cardboard")
  }
  scope :eco_choice, -> {
    where(brand: "keepcup", reusable: true)
  }
end
```

## Why not?

> If you’re working on your own, or in the early “messy” stages of an application this is probably overkill.

Reviewer: *early, "messy", stages...application, this is probably overkill.

> In a non-Rails app, importing the whole of Active Support might also be a lot of library, for a sliver of its functionality.

Reviewer: No comma between 'library' and 'for', I think...

> The principle of deprecation over time is still a good idea though, if you're working with other folks.

Reviewer: No comma between 'though' and 'if'.

---
title: "Use Active Support in Rails for deprecation messages"
description: "An internal Rails tool, availible for us"
layout: article
category: ruby
image:
  base: '2019/use-active-support-in-rails-for-deprecation-messages'
  alt: 'Caution on the pavement'
  credit: Goh Rhy Yan
  source: "https://unsplash.com/photos/FFgcWvplwsc"

---

> When authors of open source libraries change major functionality they often adhere to something called [semantic versioning](https://semver.org). At its simplest you can think of it as a rule meaning features aren’t removed until a major version is released.

Reviewer: But not necessarily...it can be the opposite: a new major version can just mean a load of new features.

So maybe say: When it comes to removing features, this isn't done until a major... or something like that?

But maybe you think the wording is fine!

Reviewer: Comma after 'major functionality' in first line.

Author: > But not necessarily...it can be the opposite: a new major version can just mean a load of new features.

Technically if it's only new, non-breaking functionality it can be a point release. Rails tend just to bundle up new features with refactors and deprecations.

Reviewer: Okay then!

Before any big changes it’s a good idea to send messages to users of the software in a harmless, but visible, way. This is often done via the logs output when the code runs.

You might have seen these when your tests run or in your live application logs.

```
blah/blah/file.rb:34: warning: constant ::Fixnum is deprecated
```

> Although in our private codebases our changes do not impact hundreds of other unseen developers, it’s still a good idea to communicate with your team when functionality changes.

Reviewer: This is reading so much better! 

I mean the whole thing until now!

The Rails framework goes through this enough that it has [a standard depreciation approach baked into Active Support](https://api.rubyonrails.org/classes/ActiveSupport/Deprecation/Reporting.html), and in _our_ code we can piggyback off the feature they use to deprecate internal features.

## Instead of...

> ...just logging

Reviewer: Colon after 'logging'

```ruby
class Coffee
  def make
    put_instant_in_cup
    add_hot_water

>     puts "We should #make_properly with an Aeropress"

Reviewer: 👍

  end

  def make_properly
    grind_on_medium_fine
    add_to_inverted_aeropress
    sleep(37)
    invert_and_apply_pressure
  end
  # ...
end
```

## Use...

...Active Support’s deprecation library.

```ruby
class Coffee
  def make
    put_instant_in_cup
    add_hot_water
    ActiveSupport::Deprecation.warn(
      "We should #make_properly with an Aeropress")
  end

  def make_properly
    grind_on_medium_fine
    add_to_inverted_aeropress
    sleep(37)
    invert_and_apply_pressure
  end
  # ...
end
```

> When you of your co-workers call `Coffee.new.make` you’ll see...

Reviewer: *one of?

Reviewer: *calls

Reviewer: Comma before "you'll"

```
DEPRECATION WARNING: We should #make_properly with an Aeropress
(called from coffee.rb:4)
```

... in your logs.

## Why?

> First, instant coffee is horrible and you _just shouldn’t_.

Reviewer: What about all those people that can't afford proper coffee...?

Wait...developers... :P

Second, this is a relatively lightweight way to 'soft protect' old ways of doing things in favour of new code.

> This is useful in two ways. One as a way to discover places in the code that use your existing method. It might be called in places you don’t expect.

Reviewer: ```suggestion
This is useful in two ways.

One is as a way to discover places in the code that use your existing method. It might be called in places you don’t expect.
```

> Another is to prevent your coworkers from adding new uses of the deprecated method because, when they attempt to, they’ll see the deprecation message.

Reviewer: *'since' instead of 'because'?

Another way to use Rails' deprecation functionality is within scopes. You'd do this when you want to maintain existing logic, but discourage further use.

```ruby
class TakeawayCoffeeCup < ApplicationRecord
  scope :good_choice, -> {
    ActiveSupport::Deprecation.warn("Be more ecologically minded")
    where(material: "cardboard")
  }
  scope :eco_choice, -> {
    where(brand: "keepcup", reusable: true)
  }
end
```

## Why not?

If you’re working on your own, or in the early, “messy”, stages of an application, this is probably overkill.

In a non-Rails app, importing the whole of Active Support might also be a lot of library for a sliver of its functionality.

> The principle of deprecation over time is still a good idea though if you're working with other folks.

Reviewer: Comma after 'though'

Author: https://github.com/andycroll/andycroll.com/pull/65#discussion_r334576162
